Build system: support alternate LIBDIR, MANDIR, DOCDIR

rworkman opened the pull request:
This addresses #926 and #927

Cyan4973 commented on programs/Makefile:
```diff
@@ -262,9 +262,7 @@ mandir      ?= $(datarootdir)/man
-MANDIR  ?= $(PREFIX)/man/man1
-else
-MANDIR  ?= $(man1dir)
+MANDIR  ?= $(PREFIX)/man
```
This construction prevents using `mandir` nor `man1dir` as an input of this Makefile.
These variables are recommended by the [GNU Makefile coding standards](https://www.gnu.org/prep/standards/html_node/Directory-Variables.html#Directory-Variables).

It also changes the destination directory for non-BSD targets, 
from `$(prefix)/share/man/man1` to `$(prefix)/man/man1`.

rworkman replied:
Yeah, this was unintentional.

Cyan4973 commented on programs/Makefile:
```diff
@@ -281,17 +279,17 @@ INSTALL_MAN     ?= $(INSTALL_DATA)
 install: zstd
 	@echo Installing binaries
-	@$(INSTALL) -d -m 755 $(DESTDIR)$(BINDIR)/ $(DESTDIR)$(MANDIR)/
+	@$(INSTALL) -d -m 755 $(DESTDIR)$(BINDIR)/ $(DESTDIR)$(MANDIR)/man1
```
Is there an issue inserting `/man1` directly into `$MANDIR` ?
Would you prefer such variable to be called `$MAN1DIR` instead ?

rworkman replied:
Disclaimer: this is my first time messing with a cmake-ish (but it's really not even that) build system; from what I recall from messing with autotools and even much older Makefiles, I've never seen MANDIR have the man? subdir in it -- it's always been $(MANDIR)/man?/

Cyan4973 commented on build/cmake/programs/CMakeLists.txt:
```diff
@@ -45,9 +45,9 @@ IF (UNIX)
-    INSTALL(FILES ${CMAKE_CURRENT_BINARY_DIR}/zstd.1 DESTINATION "share/man/man1")
-    INSTALL(FILES ${CMAKE_CURRENT_BINARY_DIR}/zstdcat.1 DESTINATION "share/man/man1")
-    INSTALL(FILES ${CMAKE_CURRENT_BINARY_DIR}/unzstd.1 DESTINATION "share/man/man1")
+    INSTALL(FILES ${CMAKE_CURRENT_BINARY_DIR}/zstd.1 DESTINATION "${CMAKE_INSTALL_MANDIR}/man1")
```
why removing `share` from the man install path ?

Cyan4973 commented on Makefile:
```diff
@@ -14,6 +14,11 @@ ZWRAPDIR = zlibWrapper
 TESTDIR  = tests
 FUZZDIR  = $(TESTDIR)/fuzz
 
+# Define some install locations
```
Why defining `cmake` variables within the root `Makefile` ?

Cyan4973 replied:
If these variables are only used by `cmake`,
maybe it's better for code maintenance to define them close to the place where they are used.

Cyan4973 commented on Makefile:
```diff
@@ -14,6 +14,11 @@ ZWRAPDIR = zlibWrapper
 
+# Define some install locations
+LIBDIR   = $(CMAKE_INSTALL_PREFIX)/lib
+MANDIR   = $(CMAKE_INSTALL_PREFIX)/share/man
```
My understanding is that `MANDIR` is different depending on Linux / BSD targets.
Here it seems there is no distinction.

Cyan4973 commented on build/cmake/programs/CMakeLists.txt:
```diff
@@ -45,9 +45,9 @@ IF (UNIX)
-    INSTALL(FILES ${CMAKE_CURRENT_BINARY_DIR}/zstdcat.1 DESTINATION "share/man/man1")
-    INSTALL(FILES ${CMAKE_CURRENT_BINARY_DIR}/unzstd.1 DESTINATION "share/man/man1")
+    INSTALL(FILES ${CMAKE_CURRENT_BINARY_DIR}/zstd.1 DESTINATION "${CMAKE_INSTALL_MANDIR}/man1")
+    INSTALL(FILES ${CMAKE_CURRENT_BINARY_DIR}/zstdcat.1 DESTINATION "${CMAKE_INSTALL_MANDIR}/man/man1")
```
I guess you meant `"${CMAKE_INSTALL_MANDIR}/man1"` ?

rworkman replied:
Indeed.

Cyan4973 commented:
How do you test the result of your installation script ?

rworkman commented:
The build process here is essentially this:

    patch -p1 < $CWD/0001-Build-system-support-alternate-LIBDIR-MANDIR-DOCDIR.patch
    
    make \
      PREFIX=/usr \
      LIBDIR=/usr/lib${LIBDIRSUFFIX} \
      MANDIR=/usr/man \
      DOCDIR=/usr/doc/$PKGNAM-$VERSION
    
    make install \
      PREFIX=/usr \
      LIBDIR=/usr/lib${LIBDIRSUFFIX} \
      MANDIR=/usr/man \
      DOCDIR=/usr/doc/$PKGNAM-$VERSION \
      DESTDIR=$PKG

About to push a new commit. I'm not convinced that it addresses the comment about where the variables are defined, but maybe it's acceptable now since they aren't specific to cmake any more.

rworkman commented:
This iteration drops DOCDIR (it wasn't used - I had planned to use it, but meh - too much trouble), add PREFIX, and addresses the review points, I think.

Cyan4973 commented:
I'm not comfortable applying this patch.
I do believe it does improve situation on your system,
but its impact seems too wide and affect different systems in hard to guess ways.

Please note that creating and maintaining a build/install script for a large number of systems is a fairly difficult exercise. While nothing is perfect, it can certainly be improved. When visibility is poor, it's important to progress cautiously by tiny steps, with limited / manageable scope.

I would be ok to start with a patch impacting only `cmake` to solve #926.
